Cache and Lookup Improvements

sdf010 opened the pull request:
added a cache to trade.py for all item mods. Also created dicts to search the items up by text and ids as this is much faster.

sdf010 committed:
edited trade.py to not call POE API every time we get all the mods

sdf010 committed:
added dicts to trade.py so you can search up item mods faster

Ethck commented on utils/trade.py:
```diff
@@ -92,6 +93,26 @@ def get_leagues() -> Tuple[str, ...]:
+        return mod_list_dict_text
+    else:
+        item_modifiers = get_item_modifiers()
+        mod_list_dict_text = {item_modifiers[i]["id"]: i for i in range(len(item_modifiers))}
```
If this is supposed to allow retrieval by text, shouldn't it be `item_modifiers[i]["text"]`? Haven't looked in a while, so feel free to correct me if I'm wrong.

sdf010 replied:
I've updated it to actually work id and text are attributes and as text is not unique it cant be used in as a dict key so we use type and text

kevr commented:
Along with this commit, we could drop in replace the tuple lookups in `parse.find_affix_match` and `parse.stat_translate` with O(1) lookups via something like the following:

```
# A crude example; preferrably done in trade.py since we've changed global access to ITEM_MODIFIERS
def stat_translate(jaffix):
  mods = get_item_modifiers_by_id()
  if jaffix in mods:
    index = mods[jaffix]
    return ITEM_MODIFIERS[index]
```

Additionally, we could add a direct lookup/return to the functions in trade.py to avoid copying the whole dictionary per lookup. That is, we could have those functions return the ItemModifier directly instead of the dictionary that contains them by passing the type, text or id to the functions themselves.

In that case, it would become something like the following to avoid copying more than we need:
```
def stat_translate(jaffix):
  return get_item_modifiers_by_id(jaffix)
```

Looks like we'll also need a mapping with keys being the types as well, in regards to how `parse.find_affix_match` looks up by types.

sdf010 committed:
corrected issues in get_items_modifiers_by_x

we can not have a text be the dict key as there are implicit and explicit which share the same text so we have to use type as part for the key.

Ethck commented on utils/trade.py:
```diff
@@ -99,7 +99,7 @@ def get_item_modifiers_by_text() -> dict:
     else:
         item_modifiers = get_item_modifiers()
-        mod_list_dict_text = {item_modifiers[i]["id"]: i for i in range(len(item_modifiers))}
+        mod_list_dict_text = {(item_modifiers[i].text, item_modifiers[i].type): item_modifiers[i] for i in range(len(item_modifiers))}
```
Awesome job. Can you please redo the lookup functions in parse.py so that we take advantage of this faster lookup utility? Seems easier to just have this PR rewrite the lookup than to break it into two parts.

kevr replied:
After some further review...

The algorithm used in find_affix_match to do the lookup on these is fundamentally different than performing a dict lookup, so it must be changed to conform with upstream in order for this patch to make any difference.

However, Df is already far into some work with reworking how we both parse and lookup items, which includes these lookups, so I'm not sure that we should change how this works in upstream at the moment.

I believe we should wait for Df's rewrites for this to be a thing.

Ethck commented:
Awaiting @sdf010 's parsing rewrite to merge. (Unless sdf010 says otherwise).
